doc/dev/error-categorization.md

```diff
@@ -0,0 +1,167 @@
+
+## Error categorization Guideline
+
+Now we will investigate each category in more detail and when to put an error/warning in there.
+
+The current structure looks like this:
+
+- Permanent errors
+  - Resource
+    - Memory Allocation
+    - General Resource
+  - Installation
+  - Logical
+    - Assertion
+    - Interface
+    - Broken plugin
+- Conflicting State
+- Timeout
+- Out-Of-Range
+- Validation
+  - Syntactic
+  - Semantic
+
+### Permanent errors ("C01000")
+
+The branch category `Permanent Errors` refer to such errors which cannot be fixed by retry
+at all. `Permanent Errors` are subdivided into
+
+- Resource
+- Installation
+- Logical
+
+#### Resource ("C01100")
+
+`Resource Errors` as a branch category are all kinds of errors which are either permission related, existence related
+or missing resources such as memory, RAM, etc.
+Examples are missing files, insufficient permissions or out of memory.
+This category forces users or applications to provide additional resources, create a missing file/ directory etc
+or provide the relevant permissions for Elektra to work.
+
+##### Memory Allocation ("C01110")
+
+`Memory Allocation Errors` are errors which come from failed `malloc` calls primarily as no
+more memory could be allocated for the application. Such errors will gain special handling
+in future releases.
+
+##### General Resource ("C01120")
+
+`General Resource Errors` are all kind of permission and existence errors. Examples are
+missing files/ directories or insufficient permission to execute certain commands (eg. you
+would require sudo permissions). Reactions are fixing the permissions or creating the file/directory
+and retry the operation.
+
+#### Installation ("C01200")
+
+`Installation Errors` are those errors which are related to a wrong installation such as
+wrong plugin names, missing backends, initialization errors, misconfiguration of Elektra etc.
+Installation errors might also be non-Elektra specific but also from dependent library/applications
+such as gpg.
+Also plugin configuration errors belong to `Installation Errors` as they this happens during
+mounting.
+Users will have to reconfigure, reinstall, recompile (with other settings) Elektra in order to
+get rid of this error or fix the installation of the corresponding library/application.
+
+#### Logical ("C01300")
+
+`Logical Errors` is another branch category in which you indicate a bug in Elektra
+such as internal errors, assertion failures or errors
+which should not happen such as going into a `default` branch when you are assured that all cases
+are covered. Usually such errors come with a message to report such failures to Elektra's bugtracker.
+Applications cannot handle such errors themselves.
+
+##### Assertion ("C01310")
+
+`Assertion Errors` are such errors which indicate a flaw in the internal logic such as going into a `default`
+branch which you do never expect to happen.
+This category might be used in the future to automatically
```

Comment:
We only describe here how it is, not what might happen in the future.

This category actually could be a sub-category of Logical.

But for now, we do not need this category, unless someone rewrites the macro ELEKTRA_ASSERT to add an error.

Reply:
It is a subcategory of Logical. Where would you then put errors such as going into a default branch despite you expect this never to happen?

Reply:
Probably "Internal" would be a better naming for assertion errors as also used in other protocols like http, grpc, etc.

Reply:
We do not need assertion errors (as our assertions do not set errors). Protocol errors are something else.

Reply:
Maybe the name assertion errors are misleading in this case then and "Internal error" is more fitting. If the category is not existent: where do you categorize errors such as [this](https://github.com/ElektraInitiative/libelektra/pull/2686/files#diff-839ad0c7bdbbc07424ca015b000030a8R1496) one?